// @preact/preset-vite/dist/esm/transform-hook-names.mjs
import { extractAssignedNames } from "@rollup/pluginutils";
import MagicString from "magic-string";
import { parseId } from "./utils.mjs";
const HOOK_IMPORTS = new Set(["preact/hooks", "preact/compat", "react"]);
const HOOK_NAME_RE = /^(useState|useReducer|useRef|useMemo)$/;
const FILTER_CODE_RE = /\buse(?:State|Reducer|Ref|Memo)\b/;
export function transformHookNamesPlugin({ devtoolsInProd, devToolsEnabled, shouldTransform, }) {
    return {
        name: "preact:transform-hook-names",
        configResolved(config) {
            devToolsEnabled =
                devToolsEnabled !== null && devToolsEnabled !== void 0 ? devToolsEnabled : (!config.isProduction || devtoolsInProd);
        },
        async transform(code, url) {
            if (!devToolsEnabled)
                return;
            const { id } = parseId(url);
            if (!shouldTransform(id))
                return;
            if (!FILTER_CODE_RE.test(code))
                return;
            let ast;
            try {
                ast = this.parse(code);
            }
            catch {
                return;
            }
            const importedHooks = getImportedHooks(ast);
            if (importedHooks.size === 0)
                return;
            const s = new MagicString(code);
            let hasHelper = false;
            const { walk } = await import("zimmerframe"); // fix for cjs until sveltejs/zimmerframe#34 is accepted
            walk(ast, null, {
                CallExpression(node, { path, next }) {
                    const callee = node.callee;
                    if (callee.type !== "Identifier") {
                        next();
                        return;
                    }
                    const hookName = callee.name;
                    // the hook name might be shadowed by a local variable,
                    // but that false-positive is fine
                    // since `addHookName` is transparent
                    if (!HOOK_NAME_RE.test(hookName) || !importedHooks.has(hookName)) {
                        next();
                        return;
                    }
                    const parent = path[path.length - 1];
                    const bindingNames = getOuterBindingNames(parent);
                    if (bindingNames.length === 0) {
                        next();
                        return;
                    }
                    let name = bindingNames[0];
                    hasHelper = true;
                    s.prependLeft(node.start, "addHookName(");
                    s.appendRight(node.end, `, ${JSON.stringify(name)})`);
                    next();
                },
            });
            if (!hasHelper)
                return;
            const firstNode = ast.body[0];
            s.prependLeft(firstNode.start, 'import { addHookName } from "preact/devtools";\n');
            return {
                code: s.toString(),
                map: s.generateMap({ hires: "boundary" }),
            };
        },
    };
}
function getImportedHooks(ast) {
    const hooks = new Set();
    for (const node of ast.body) {
        if (node.type !== "ImportDeclaration")
            continue;
        const source = typeof node.source.value === "string" ? node.source.value : null;
        if (!source || !HOOK_IMPORTS.has(source))
            continue;
        for (const specifier of node.specifiers) {
            if (specifier.type !== "ImportSpecifier")
                continue;
            const importedName = specifier.imported.name;
            if (!HOOK_NAME_RE.test(importedName) ||
                specifier.local.name !== importedName)
                continue;
            hooks.add(importedName);
        }
    }
    return hooks;
}
function getOuterBindingNames(node) {
    if (!node)
        return [];
    switch (node.type) {
        case "VariableDeclarator":
            return extractAssignedNames(node.id);
        case "AssignmentExpression":
            return extractAssignedNames(node.left);
        case "Identifier":
        case "ArrayPattern":
        case "ObjectPattern":
        case "RestElement":
        case "AssignmentPattern":
            return extractAssignedNames(node);
        default:
            return [];
    }
}
