// @preact/preset-vite/dist/cjs/transform-hook-names.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformHookNamesPlugin = void 0;
const pluginutils_1 = require("@rollup/pluginutils");
const magic_string_1 = __importDefault(require("magic-string"));
const utils_js_1 = require("./utils.js");
const HOOK_IMPORTS = new Set(["preact/hooks", "preact/compat", "react"]);
const HOOK_NAME_RE = /^(useState|useReducer|useRef|useMemo)$/;
const FILTER_CODE_RE = /\buse(?:State|Reducer|Ref|Memo)\b/;
function transformHookNamesPlugin({ devtoolsInProd, devToolsEnabled, shouldTransform, }) {
    return {
        name: "preact:transform-hook-names",
        configResolved(config) {
            devToolsEnabled =
                devToolsEnabled !== null && devToolsEnabled !== void 0 ? devToolsEnabled : (!config.isProduction || devtoolsInProd);
        },
        async transform(code, url) {
            if (!devToolsEnabled)
                return;
            const { id } = (0, utils_js_1.parseId)(url);
            if (!shouldTransform(id))
                return;
            if (!FILTER_CODE_RE.test(code))
                return;
            let ast;
            try {
                ast = this.parse(code);
            }
            catch {
                return;
            }
            const importedHooks = getImportedHooks(ast);
            if (importedHooks.size === 0)
                return;
            const s = new magic_string_1.default(code);
            let hasHelper = false;
            const { walk } = await Promise.resolve().then(() => __importStar(require("zimmerframe"))); // fix for cjs until sveltejs/zimmerframe#34 is accepted
            walk(ast, null, {
                CallExpression(node, { path, next }) {
                    const callee = node.callee;
                    if (callee.type !== "Identifier") {
                        next();
                        return;
                    }
                    const hookName = callee.name;
                    // the hook name might be shadowed by a local variable,
                    // but that false-positive is fine
                    // since `addHookName` is transparent
                    if (!HOOK_NAME_RE.test(hookName) || !importedHooks.has(hookName)) {
                        next();
                        return;
                    }
                    const parent = path[path.length - 1];
                    const bindingNames = getOuterBindingNames(parent);
                    if (bindingNames.length === 0) {
                        next();
                        return;
                    }
                    let name = bindingNames[0];
                    hasHelper = true;
                    s.prependLeft(node.start, "addHookName(");
                    s.appendRight(node.end, `, ${JSON.stringify(name)})`);
                    next();
                },
            });
            if (!hasHelper)
                return;
            const firstNode = ast.body[0];
            s.prependLeft(firstNode.start, 'import { addHookName } from "preact/devtools";\n');
            return {
                code: s.toString(),
                map: s.generateMap({ hires: "boundary" }),
            };
        },
    };
}
exports.transformHookNamesPlugin = transformHookNamesPlugin;
function getImportedHooks(ast) {
    const hooks = new Set();
    for (const node of ast.body) {
        if (node.type !== "ImportDeclaration")
            continue;
        const source = typeof node.source.value === "string" ? node.source.value : null;
        if (!source || !HOOK_IMPORTS.has(source))
            continue;
        for (const specifier of node.specifiers) {
            if (specifier.type !== "ImportSpecifier")
                continue;
            const importedName = specifier.imported.name;
            if (!HOOK_NAME_RE.test(importedName) ||
                specifier.local.name !== importedName)
                continue;
            hooks.add(importedName);
        }
    }
    return hooks;
}
function getOuterBindingNames(node) {
    if (!node)
        return [];
    switch (node.type) {
        case "VariableDeclarator":
            return (0, pluginutils_1.extractAssignedNames)(node.id);
        case "AssignmentExpression":
            return (0, pluginutils_1.extractAssignedNames)(node.left);
        case "Identifier":
        case "ArrayPattern":
        case "ObjectPattern":
        case "RestElement":
        case "AssignmentPattern":
            return (0, pluginutils_1.extractAssignedNames)(node);
        default:
            return [];
    }
}
